daemon/cmd/devices.go

```diff
@@ -100,36 +98,40 @@ func (dm *DeviceManager) Detect() error {
 		detectIPv6MCastDev = false
 	}
 
-	k8sNodeDev := ""
-	k8sNodeLink, err := findK8SNodeIPLink()
-	if err == nil {
-		k8sNodeDev = k8sNodeLink.Attrs().Name
-		dm.devices[k8sNodeDev] = struct{}{}
-	}
+	if detectDirectRoutingDev || detectIPv6MCastDev {
+		k8sNodeDev := ""
+		k8sNodeLink, err := findK8SNodeIPLink()
+		if err == nil {
+			k8sNodeDev = k8sNodeLink.Attrs().Name
+			dm.devices[k8sNodeDev] = struct{}{}
+		} else if k8s.IsEnabled() {
+			return fmt.Errorf("k8s is enabled, but still failed to find node IP: %w", err)
+		}
 
-	if detectDirectRoutingDev {
-		// If only one device found, use that one. Otherwise use the device with k8s node IP.
-		if len(dm.devices) == 1 {
-			for dev := range dm.devices {
-				option.Config.DirectRoutingDevice = dev
-				break
+		if detectDirectRoutingDev {
+			// If only one device found, use that one. Otherwise use the device with k8s node IP.
+			if len(dm.devices) == 1 {
+				for dev := range dm.devices {
+					option.Config.DirectRoutingDevice = dev
+					break
+				}
+			} else if k8sNodeDev != "" {
+				option.Config.DirectRoutingDevice = k8sNodeDev
+			} else {
+				return fmt.Errorf("Unable to determine direct routing device. Use --%s to specify it",
+					option.DirectRoutingDevice)
 			}
-		} else if k8sNodeDev != "" {
-			option.Config.DirectRoutingDevice = k8sNodeDev
-		} else {
-			return fmt.Errorf("Unable to determine direct routing device. Use --%s to specify it",
-				option.DirectRoutingDevice)
+			log.Infof("Detected %s=%s", option.DirectRoutingDevice, option.Config.DirectRoutingDevice)
```

Comment:
While you are here, maybe you could fix the msg, so that we respect the structured logging?

Reply:
Sure, good catch.